File: apps/api/internal/domain/running/running_test.go
package running

import (
	"encoding/json"
	"testing"

	"github.com/hamid-karimi/coachin/apps/api/internal/golden"
)

type tc struct {
	Fn       string          `json:"fn"`
	D1       float64         `json:"d1"`
	T1       float64         `json:"t1"`
	D2       float64         `json:"d2"`
	Value    json.RawMessage `json:"value"`
	Seconds  float64         `json:"seconds"`
	Target   string          `json:"target"`
	CustomKm *float64        `json:"customKm"`
	PBs      PBs             `json:"pbs"`
	TargetKm float64         `json:"targetKm"`
	Want     json.RawMessage `json:"want"`
}

func TestRunningMatchesLegacy(t *testing.T) {
	run := map[string]func(t *testing.T, c tc) any{
		"riegelSeconds": func(_ *testing.T, c tc) any { return RiegelSeconds(c.D1, c.T1, c.D2) },
		"parseTimeToSeconds": func(t *testing.T, c tc) any {
			return golden.OrNull(ParseTimeToSeconds(golden.Decode[string](t, c.Value)))
		},
		"formatSeconds": func(_ *testing.T, c tc) any { return FormatSeconds(c.Seconds) },
		"clampBaseWeeks": func(t *testing.T, c tc) any {
			return ClampBaseWeeks(golden.Decode[*float64](t, c.Value))
		},
		"raceDistanceKm": func(_ *testing.T, c tc) any {
			return golden.OrNull(RaceDistanceKm(c.Target, c.CustomKm))
		},
		"suggestGoalForDistance": func(_ *testing.T, c tc) any {
			return golden.OrNull(SuggestGoalForDistance(c.PBs, c.TargetKm))
		},
	}
	for _, c := range golden.Load[tc](t, "running") {
		fn, ok := run[c.Fn]
		if !ok {
			t.Fatalf("no Go port for %q", c.Fn)
		}
		if equal, got := golden.Equal(t, fn(t, c), c.Want); !equal {
			t.Errorf("%s(%+v) = %s, want %s", c.Fn, c, got, c.Want)
		}
	}
}

File: apps/web/lib/running.test.ts
import { readFileSync } from "node:fs";
import path from "node:path";
import { describe, expect, it } from "vitest";
import {
  clampBaseWeeks,
  formatSeconds,
  parseTimeToSeconds,
  raceDistanceKm,
  riegelSeconds,
  suggestGoalForDistance,
  type RaceTarget,
} from "./running";

// The API owns this math (apps/api/internal/domain/running); the wizard keeps
// a copy for instant goal suggestions. Replaying the same golden vectors as
// the Go tests keeps the two identical.
type Case = Record<string, unknown> & { fn: string; want: unknown };
const vectors: Case[] = JSON.parse(
  readFileSync(path.resolve(__dirname, "../../../testdata/golden/running.json"), "utf8"),
).cases;

const RUN: Record<string, (c: Case) => unknown> = {
  riegelSeconds: (c) => riegelSeconds(c.d1 as number, c.t1 as number, c.d2 as number),
  parseTimeToSeconds: (c) => parseTimeToSeconds(c.value as string),
  formatSeconds: (c) => formatSeconds(c.seconds as number),
  raceDistanceKm: (c) => raceDistanceKm(c.target as RaceTarget, c.customKm as number | null),
  clampBaseWeeks: (c) => clampBaseWeeks(c.value as never),
  suggestGoalForDistance: (c) => suggestGoalForDistance(c.pbs as never, c.targetKm as number),
};

describe("running (golden vectors shared with the Go API)", () => {
  it("covers every function", () => {
    expect(new Set(vectors.map((c) => c.fn))).toEqual(new Set(Object.keys(RUN)));
  });

  it.each(vectors.map((c, i) => [i, c] as const))("case %i", (_, c) => {
    expect(RUN[c.fn](c)).toEqual(c.want);
  });
});
